lib/cn/spill.c

```diff
@@ -180,394 +180,437 @@ kv_spill_abort(struct cn_tstate_omf *omf, void *arg)
-                &vdata, &vlen, &complen))
-            break;
 
-        if (vtype == vtype_val)
-            omlen = vlen;
-        else if (vtype == vtype_cval)
-            omlen = complen;
-        else
-            omlen = 0;
-
-        direct = omlen > direct_read_len;
-
-        /* [HSE_REVISIT] direct read path allocates buffer. Performing
-         * direct reads into the buffer in kvset builder without this
-         * is a future opportunity.
-         */
-        if (direct) {
-            uint bufsz_min;
-
-            if (ev(omlen > HSE_KVS_VALUE_LEN_MAX)) {
-                assert(omlen <= HSE_KVS_VALUE_LEN_MAX);
-                err = merr(EBUG);
-                goto done;
-            }
+    return cnum;
+}
 
-            bufsz_min = omlen;
+static void
+put_route(struct cn_compaction_work *w, uint cnum)
+{
+    if (w->cw_action == CN_ACTION_SPILL)
+        cn_tree_route_put(w->cw_tree, cnum);
+}
 
-            /* If value offset is not page aligned then allocate
-             * one additional page to prevent a potential copy
-             * inside direct read function
-             */
-            if (vboff % PAGE_SIZE)
-                bufsz_min += PAGE_SIZE;
+/**
+ * cn_spill() - merge key-value streams, then spill kv-pairs one node at a time or
+ *              kv-compact into a node.
+ * Requirements:
+ *   - Each input iterator must produce keys in sorted order.
+ *   - Iterator iterv[i] must contain newer entries than iterv[i+1].
+ */
+merr_t
+cn_spill(struct cn_compaction_work *w)
+{
+    struct bin_heap *bh;
+    struct merge_item curr;
+    struct kvset_builder *child = NULL;
+    struct cn_khashmap *khashmap = NULL;
+    struct key_obj ekobj = { 0 }, prev_kobj;
+    struct cn_tree *tree = w->cw_tree;
 
-            if (!buf || bufsz < bufsz_min) {
-                free_aligned(buf);
+    char ekbuf[HSE_KVS_KEY_LEN_MAX];
+    uint vlen, complen, omlen, eklen, direct_read_len, cnum, prev_cnum;
+    uint curr_klen HSE_MAYBE_UNUSED;
+    u32 bufsz = 0;
+    void *buf = NULL;
+    merr_t err;
 
-                if (omlen < HSE_KVS_VALUE_LEN_MAX / 4)
-                    bufsz = HSE_KVS_VALUE_LEN_MAX / 4;
-                else if (omlen < HSE_KVS_VALUE_LEN_MAX / 2)
-                    bufsz = HSE_KVS_VALUE_LEN_MAX / 2;
-                else
-                    bufsz = HSE_KVS_VALUE_LEN_MAX;
+    u64  seq, emitted_seq = 0, emitted_seq_pt = 0;
+    bool emitted_val = false, bg_val = false, more, gt_max_edge = false;
+    bool kvcompact = (w->cw_action == CN_ACTION_COMPACT_KV);
+    bool spill = (w->cw_action == CN_ACTION_SPILL);
 
-                /* add an extra page if not aligned */
-                if (bufsz_min < bufsz)
-                    bufsz += PAGE_SIZE;
+    u64  tstart, tprog = 0;
+    u64  dbg_prev_seq HSE_MAYBE_UNUSED;
+    uint dbg_prev_src HSE_MAYBE_UNUSED;
+    uint dbg_nvals_this_key HSE_MAYBE_UNUSED;
+    uint seqno_errcnt = 0;
+    bool dbg_dup HSE_MAYBE_UNUSED;
```

Comment:
Kind of like the 'dbg' vars together.

Reply:
Sure, I can swap seqno_errcnt and dbg_dup. Btw, seqno_errcnt is also used for debugging.